File: 002_Java/20190831_2235_checkStringPermutation/isStringPermutation_v01.java


/*
Given 2 strings, write a method to decide if one is the permutation of the other

*/
import java.util.Arrays; 

class IsPermutation {

	static final int MAX = 256;

	static public boolean isPermute(String a, String b){
		int[] charTable = new int[MAX]; /* it will by default set all values to 0 */
		int sizeA = a.length();
		int sizeB = b.length();
		if((sizeA==0)||(sizeB==0)||(sizeA!=sizeB)||(a.equals(b))){
			return(false);
		}
		for(int i=0;i<sizeA;i++){
			charTable[(int)a.charAt(i)]++;
		}
		for(int i=0;i<sizeA;i++){
			if ((charTable[(int)b.charAt(i)]-=1)<0){
				return(false);
			}
		}

		return(true);
	}

	// using toCharArray()
	static public boolean isPermuteSorted(String a, String b){
		int sizeA = a.length();
		int sizeB = b.length();
		char[] a_array = a.toCharArray();
		char[] b_array = b.toCharArray();

		if((sizeA==0)||(sizeB==0)||(sizeA!=sizeB)||(a.equals(b))){
			return(false);
		}
		
		Arrays.sort(a_array);
		Arrays.sort(b_array);
		if(!a_array.equals(b_array)){return(false);}

		return(true);
	}

	static public void main(String[] args){
		String a = "implement";
		String b = "tnemelpmi";
		String c = "slip";
		String d = "tneme0pmi";

		System.out.print( a + " and " + a + " are");
		if (!isPermute(a,a)){
			System.out.print(" not");
		}
		System.out.println( " permuted");

		System.out.print( a + " and " + b + " are");
		if (!isPermute(a,b)){
			System.out.print(" not");
		}
		System.out.println( " permuted");

		System.out.print( a + " and " + c + " are");
		if (!isPermute(a,c)){
			System.out.print(" not");
		}
		System.out.println( " permuted");

		System.out.print( a + " and " + d + " are");
		if (!isPermute(a,d)){
			System.out.print(" not");
		}
		System.out.println( " permuted");

		///////////////////////////////////////////////////////////

		System.out.print( a + " and " + a + " are");
		if (!isPermuteSorted(a,a)){
			System.out.print(" not");
		}
		System.out.println( " permuted");

		System.out.print( a + " and " + b + " are");
		if (!isPermuteSorted(a,b)){
			System.out.print(" not");
		}
		System.out.println( " permuted");

		System.out.print( a + " and " + c + " are");
		if (!isPermuteSorted(a,c)){
			System.out.print(" not");
		}
		System.out.println( " permuted");

		System.out.print( a + " and " + d + " are");
		if (!isPermuteSorted(a,d)){
			System.out.print(" not");
		}
		System.out.println( " permuted");


	}
}


/*a = "implement"
b = "tnemelpmi"
c = "slip"
d = "tneme0pmi"

javac isTringPermutation_v01.java
java IsPermutation
*/

File: 001_Python/20190831_2235_checkStringPermutation/isStringPermutation_v01.py
# coding=utf-8

'''
Given 2 strings, write a method to decide if one is the permutation of the other

'''
def isStringsPermutedDic0(a,b) :
	if len(a) != len(b) or len(a) == 0 or len(b) == 0 or a == b :
		return(False)
	dictio = {}
	for c in a :
		if c in dictio :
			dictio[c] += 1
		else :
			dictio[c] = 1
	# here hash table for string a is computed.
	# Now we can decrement with chars in b string
	for c in b :
		if c in dictio :
			dictio[c] -= 1
		else :
			return(False)
	# at the end, if all values in diction are 0 --> string permuted, otherwise, not permuted.
	for c in dictio :
		if dictio[c] != 0 :
			return(False)
	# if code reached this step --> all dictio[c] where at 0 --> permuted strings --> return(True)
	return(True)

def isStringsPermutedDic1(a,b) :
	if len(a) != len(b) or len(a) == 0 or len(b) == 0 or a == b :
		return(False)
	dictio = {}
	for c in a :
		if c in dictio :
			dictio[c] += 1
		else :
			dictio[c] = 1
	# here hash table for string a is computed.
	# Build hash table for string b
	dictio_b = {}
	for c in b :
		if c in dictio_b :
			dictio_b[c] += 1
		else :
			dictio_b[c] = 1
	# here hash table for string b is computed.
	# now we can return comparison of dictionaries
	#return(cmp(dictio,dictio_b))
	# however cmp() does not exist anymore in Python 3.
	# so we need to compare manually the dictionnaries.
	return(sorted(dictio) == sorted(dictio_b))

def isStringsPermutedSort(a,b) :
	if len(a) != len(b) or len(a) == 0 or len(b) == 0 or a == b :
		return(False)
	return(sorted(a) == sorted(b))



a = "implement"
b = "tnemelpmi"
c = "slip"
d = "tneme0pmi"

print("using isStringsPermutedDic0")
print( a + " and " + a )
if isStringsPermutedDic0(a,a) : print("True")
else : print("False")
print( a + " and " + b )
if isStringsPermutedDic0(a,b) : print("True")
else : print("False")
print( a + " and " + c )
if isStringsPermutedDic0(a,c) : print("True")
else : print("False")
print( a + " and " + d )
if isStringsPermutedDic0(a,d) : print("True")
else : print("False")

print("using isStringsPermutedDic1")
print( a + " and " + a )
if isStringsPermutedDic1(a,a) : print("True")
else : print("False")
print( a + " and " + b )
if isStringsPermutedDic1(a,b) : print("True")
else : print("False")
print( a + " and " + c )
if isStringsPermutedDic1(a,c) : print("True")
else : print("False")
print( a + " and " + d )
if isStringsPermutedDic1(a,d) : print("True")
else : print("False")

print("using isStringsPermutedSort")
print( a + " and " + a )
if isStringsPermutedSort(a,a) : print("True")
else : print("False")
print( a + " and " + b )
if isStringsPermutedSort(a,b) : print("True")
else : print("False")
print( a + " and " + c )
if isStringsPermutedSort(a,c) : print("True")
else : print("False")
print( a + " and " + d )
if isStringsPermutedSort(a,d) : print("True")
else : print("False")

'''
python3 isStringPermutaiton_v01.py 

'''


